contrib/mesos/pkg/queue/delay/indexed_delay_queue.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package delay
+
+import (
+	"container/heap"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pivotal-golang/clock"
+
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue"
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue/priority"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// IndexedDelayQueue is a thread-safe, time-based queue with uniquely identified items.
+//
+// Adding or offering an item with an ID matching an item still in the queue will
+// invoke the configured SchedulingPolicy and supplied ValueReplacementPolicy to
+// determine the desired behavior.
+//
+// Items can only be popped after their event time has been reached.
+// Use Add to specify a delay duration and have the event time calculated.
+// Use Offer to specify a specific event time.
+type IndexedDelayQueue struct {
+	*Queue
```

Comment:
what do you think about moving Queue to queue/internal/queue.go? that takes
advantage of go's "internal" protections and only allows access to Queue
from things in the queue/ subtree.

On Mon, Sep 28, 2015 at 9:43 PM, Karl Isenberg <notifications@github.com>
wrote:

> In contrib/mesos/pkg/queue/delay/indexed_delay_queue.go
> <https://github.com/kubernetes/kubernetes/pull/14610#discussion_r40626517>
> :
>
> > +	"k8s.io/kubernetes/contrib/mesos/pkg/queue"
> > +	"k8s.io/kubernetes/contrib/mesos/pkg/queue/priority"
> > +	"k8s.io/kubernetes/pkg/util/sets"
> > +)
> > +
> > +// IndexedDelayQueue is a thread-safe, time-based queue with uniquely identified items.
> > +//
> > +// Adding or offering an item with an ID matching an item still in the queue will
> > +// invoke the configured SchedulingPolicy and supplied ValueReplacementPolicy to
> > +// determine the desired behavior.
> > +//
> > +// Items can only be popped after their event time has been reached.
> > +// Use Add to specify a delay duration and have the event time calculated.
> > +// Use Offer to specify a specific event time.
> > +type IndexedDelayQueue struct {
> > +	*Queue
>
> Having a named field is an option, but it means aliasing all the delegate
> functions, which I'm not sure is worth it. Plus it makes the class less
> extensible. IndexedDelayQueue, for example, reaches in to Queue.Queue. it
> wouldn't be able to do that if it Queue were private and IndexedDelayQueue
> was in another package.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14610/files#r40626517>.
>